### ISSUE TITLE
[Bug]: `aws_grafana_role_association` resource has unexpected/dangerous behaviour with `for_each` or `count`

### ISSUE THREAD

**Issue by RuairiSinkler** [hashicorp/terraform-provider-aws#32510]
### Terraform Core Version

1.2.9

### AWS Provider Version

3.59.0

### Affected Resource(s)

`aws_grafana_role_association`

### Expected Behavior

When using `for_each` to manage users independently with the `aws_grafana_role_association`, each resource acts
independently, managing the association between that user and role as a single unit.

### Actual Behavior

Each "independent" resource refers to the same actual resource/role association, causing each one to be a duplicate of the other.

The other strange and dangerous side effect of this is that when first applying this sort of setup, the `plan` implies that each resource will be independent, but only after applying do they all become duplicates. This also means the terraform setup can become locked into an unchangeable state if >50 users are added at once.

### Relevant Error/Panic Output Snippet

```shell
│ Error: error creating Grafana Workspace Role Association: ValidationException: Total instruction count cannot be > 50
│ {
│   RespMetadata: {
│     StatusCode: 400,
│     RequestID: "..."
│   },
│   Message_: "Total instruction count cannot be > 50"
│ }
```

### Terraform Configuration Files

With a Grafana workspace, and a list of user data in `local.users` with length > 1:

```terraform
resource "aws_grafana_role_association" "viewer" {
  for_each = local.users

  role         = "VIEWER"
  user_ids     = [each.value.id]
  workspace_id = aws_grafana_workspace.grafana.id
}
```

### Steps to Reproduce

With a list of users, try to plan and apply the above role association.

Upon planning, the output will show independent resources for each user, each with a single user id in the association data e.g. with 2 users (anonymised as user1 and user2):

```terraform
  # module...aws_grafana_role_association.viewer[user1] will be created
  + resource "aws_grafana_role_association" "viewer" {
      + id           = (known after apply)
      + role         = "VIEWER"
      + user_ids     = [
          + user1.id,
        ]
      + workspace_id = ...
    }

  # module...aws_grafana_role_association.viewer[user2] will be created
  + resource "aws_grafana_role_association" "viewer" {
      + id           = (known after apply)
      + role         = "VIEWER"
      + user_ids     = [
          + user2.id,
        ]
      + workspace_id = ...
    }

```

Then, applying that and planning again **without any changes** results in showing each user independently, but trying to delete the other user's id from their association reference since they refer to every user each:
```terraform
  # module...aws_grafana_role_association.viewer[user1] will be updated in-place
  ~ resource "aws_grafana_role_association" "viewer" {
        id           = "g-f69c56d6b6/VIEWER"
      ~ user_ids     = [
          - user2.id,
            # (1 unchanged element hidden)
        ]
        # (3 unchanged attributes hidden)
    }

  # module...aws_grafana_role_association.viewer[user2] will be updated in-place
  ~ resource "aws_grafana_role_association" "viewer" {
        id           = "g-f69c56d6b6/VIEWER"
      ~ user_ids     = [
          - user1.id,
            # (1 unchanged element hidden)
        ]
        # (3 unchanged attributes hidden)
    }

```

However, adding or removing a user and performing a plan will further show each resource in fact refers to every user available e.g. removing user1 from the above:

```terraform
  # module...aws_grafana_role_association.viewer[user1] will be destroyed
  # (because key [user1] is not in for_each map)
  - resource "aws_grafana_role_association" "viewer" {
      - group_ids    = [] -> null
      - id           = "g-f69c56d6b6/VIEWER" -> null
      - role         = "VIEWER" -> null
      - user_ids     = [
          - user1.id,
          - user2.id,
        ] -> null
      - workspace_id = ... -> null
    }

  # module...aws_grafana_role_association.viewer[user2] will be updated in-place
  ~ resource "aws_grafana_role_association" "viewer" {
        id           = "g-f69c56d6b6/VIEWER"
      ~ user_ids     = [
          - user1.id,
            # (1 unchanged element hidden)
        ]
        # (3 unchanged attributes hidden)
    }

```

Applying the above seems like it would produce inconsistent results based on the order of users too, since they are both trying to manage the state of the same role association and so ordering them differently may result in either just the one user being left, or the role association being gone altogether.

The major problem here is that if this is performed with >50 users, the terraform state becomes locked, and can only be resolved through manual manipulation of the terraform state directly. This is also the reason we came across this bug, as we were trying to work around the 50 user limit when providing a list to the `user_ids` parameter.

i.e. with a list of 51 users, creating a `aws_grafana_role_association` `for_each` of them, associating the role with their ID alone in the resource, results in a plan that lists each one independently, and which applies without issue

However, the resulting state is actually a single role associated with each user - this makes sense - but does mean that each of the 51 independent resources all refer to the same backing role association, all with the 51 users inside them. This means trying to add or delete even a single user attempts to delete the resource referring to 51 users, and results in a `Message_: "Total instruction count cannot be > 50"` error. Thus rendering the terraform state locked and unchangeable.

### Important Factoids

The reason we came across this issue in the first place is the limit of 50 users allowed when supplying user_ids to `aws_grafana_role_association` - using groups is a difficult option due to our SSO setup meaning managing users directly and independently is the preferable option (though this is not the case, as this has proven this will be an impossibility).

### Would you like to implement a fix?

No
